java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,336 +21,312 @@
+        if (result.isDetermined()) {
+          return result;
+        }
+        usedRelations.add(relation);
+      }
+      relations = deduce(usedRelations);
+    }
+    return RelationState.UNDETERMINED;
   }
 
+  static Set<BinaryRelation> deduce(Set<BinaryRelation> relations) {
+    Set<BinaryRelation> result = new HashSet<>();
+    BinaryRelation[] binaryRelations = relations.toArray(new BinaryRelation[relations.size()]);
+    for (int i = 0; i < binaryRelations.length; i++) {
+      for (int j = i + 1; j < binaryRelations.length; j++) {
+        BinaryRelation a = binaryRelations[i];
+        BinaryRelation b = binaryRelations[j];
+        BinaryRelation simplified = simplify(a, b);
+        if (simplified != null && !relations.contains(simplified)) {
+          result.add(simplified);
+        }
+        if (a.inTransitiveRelationship(b)) {
+          BinaryRelation transitive = combineTransitively(a, b);
+          if (transitive != null && !relations.contains(transitive)) {
+            result.add(transitive);
+          }
+        }
+      }
+    }
+    return result;
+  }
 
-  public static class TransitiveRelationExceededException extends RuntimeException {
-    public TransitiveRelationExceededException() {
-      super("Number of transitive relations exceeded!");
+  @CheckForNull
+  private static BinaryRelation simplify(BinaryRelation a, BinaryRelation b) {
+    // a >= b && b >= a -> a == b
+    if (a.kind == GREATER_THAN_OR_EQUAL && b.kind == GREATER_THAN_OR_EQUAL
+      && a.hasSameOperandsAs(b) && !a.equals(b)) {
+      return binaryRelation(EQUAL, a.leftOp, a.rightOp);
     }
+    return null;
   }
 
-  @Override
-  public int hashCode() {
-    return hashcode;
+  private boolean hasSameOperand() {
+    return leftOp.equals(rightOp);
   }
 
-  @Override
-  public boolean equals(Object obj) {
-    if (obj instanceof BinaryRelation) {
-      return equalsRelation((BinaryRelation) obj);
-    }
-    return false;
+  boolean hasOperand(SymbolicValue operand) {
+    return leftOp.equals(operand) || rightOp.equals(operand);
   }
 
-  private boolean equalsRelation(BinaryRelation rel) {
-    if (kind.equals(rel.kind)) {
-      return leftOp.id() == rel.leftOp.id() && rightOp.id() == rel.rightOp.id();
-    }
-    return false;
+  private boolean hasSameOperandsAs(BinaryRelation other) {
+    return (leftOp.equals(other.leftOp) && rightOp.equals(other.rightOp))
+      || (leftOp.equals(other.rightOp) && rightOp.equals(other.leftOp));
   }
 
-  @Override
-  public String toString() {
-    StringBuilder buffer = new StringBuilder();
-    buffer.append(leftOp);
-    buffer.append(kind.operand);
-    buffer.append(rightOp);
-    return buffer.toString();
+  @VisibleForTesting
+  SymbolicValue differentOperand(BinaryRelation other) {
+    // FIXME remove this sanity check
+    Preconditions.checkState(inTransitiveRelationship(other), toString() + " is not in transitive relationship with " + other);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Invoke method(s) only conditionally. Use the built-in formatting to construct this argument. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)